php_phongo.c

```diff
@@ -3751,10 +3739,42 @@ PHP_MINIT_FUNCTION(mongodb)
 /* {{{ PHP_MSHUTDOWN_FUNCTION */
 PHP_MSHUTDOWN_FUNCTION(mongodb)
 {
+	HashTable* pclients = &MONGODB_G(pclients);
 	(void) type; /* We don't care if we are loaded via dl() or extension= */
 
-	/* Destroy HashTable for persistent clients. The HashTable destructor will
-	 * destroy any mongoc_client_t objects that were created by this process. */
+	/* Destroy mongoc_client_t objects in reverse order. This is necessary to
+	 * prevent segmentation faults as clients may reference other clients in
+	 * encryption settings. */
+#if PHP_VERSION_ID >= 70000
+	{
+		zval* z_ptr;
+
+		ZEND_HASH_REVERSE_FOREACH_VAL(pclients, z_ptr)
```

Comment:
I didn't realize macros and functions for reverse iteration existed. That's splendid!